### PR TITLE
Add FocalLoss, class weights, and enhanced tile filtering

### LINKED ISSUE #335: Geoai Improvement to Support Training Using Discrete Class Landcover Data
### Description

Hello geoai community. I would like to propose some additional parameters I have added to the Geoai package to address the issues I had with trying to train a landcover model where my training data is sparse, instead of a continuous landcover raster (as in the provided example). My discrete landcover classification has severe class imbalance (e.g., roads <1% vs forests >40% of pixels) and needed a flexible loss functions that can handle sparse, categorical geospatial data.
These changes have improved my use purpose greatly.

🚀 Proposed Enhancements Added:

1. New Loss Function
FocalLoss: Custom PyTorch implementation specifically for class imbalance
Purpose: Focuses training on "hard examples" while down-weighting easy predictions
Benefit: Prevents dominant classes (forests) from overwhelming rare classes (roads, wetlands)
Parameters: [focal_alpha](class balancing) and [focal_gamma](hard example focus)

2. Class Weight System
Dual-mode operation:
Inverse Frequency Mode: Auto-computes weights based on pixel abundance
Pure Custom Mode: Manual weight control without frequency calculation
Custom multipliers: Fine-tune specific landcover classes if one is being overpredicted
Weight capping: Prevents extreme values that destabilize training
Purpose: Ensures rare landcover types get appropriate attention during training

3. Enhanced Tile Filtering
min_feature_ratio: Filters out background-heavy tiles
Purpose: Improves training efficiency by focusing on tiles with meaningful habitat content
Benefit: Reduces class imbalance at the tile level, leading to better feature learning

4. Flexible Ignore Index
Configurable pixel exclusion: Handle unlabeled/uncertain areas
Purpose: Skip pixels with ambiguous or missing habitat labels
Benefit: Prevents model from learning incorrect patterns from uncertain data

### Source code

```
1. New Loss Function

class FocalLoss(torch.nn.Module):
    """
    Focal Loss for addressing class imbalance in segmentation.
    
    Reference: Lin, T. Y., Goyal, P., Girshick, R., He, K., & Dollár, P. (2017).
    Focal loss for dense object detection. ICCV.
    """
    
    def __init__(self, alpha=1.0, gamma=2.0, ignore_index=-100, reduction='mean', weight=None):
        super(FocalLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.ignore_index = ignore_index
        self.reduction = reduction
        self.weight = weight  # Class weights tensor
        
    def forward(self, inputs, targets):
        """
        Args:
            inputs: Tensor of shape [N, C, H, W] where C is number of classes
            targets: Tensor of shape [N, H, W] with class indices
        """
        # Handle ignore_index parameter - if False, disable ignoring
        if self.ignore_index is False:
            ignore_idx = -100  # Use a value that won't match any target
        else:
            ignore_idx = self.ignore_index
        
        # Apply log_softmax to get log probabilities
        log_pt = F.log_softmax(inputs, dim=1)
        
        # Use NLL loss with ignore_index and class weights to handle ignored pixels properly
        ce_loss = F.nll_loss(log_pt, targets, ignore_index=ignore_idx, weight=self.weight, reduction='none')
        
        # Get probabilities by exponentiating the negative cross entropy
        pt = torch.exp(-ce_loss)
        
        # Apply focal loss formula: FL = -alpha * (1-pt)^gamma * log(pt)
        focal_loss = self.alpha * (1 - pt) ** self.gamma * ce_loss
        
        # Handle reduction
        if self.reduction == 'mean':
            # Only average over non-ignored pixels
            if ignore_idx != -100:
                valid_mask = (targets != ignore_idx)
                if valid_mask.sum() > 0:
                    return focal_loss[valid_mask].mean()
                else:
                    return torch.tensor(0.0, device=inputs.device, requires_grad=True)
            else:
                return focal_loss.mean()
        elif self.reduction == 'sum':
            if ignore_idx != -100:
                valid_mask = (targets != ignore_idx)
                return focal_loss[valid_mask].sum()
            else:
                return focal_loss.sum()
        else:
            return focal_loss


def get_loss_function(loss_name: str, 
                     ignore_index: Union[int, bool] = -100, 
                     num_classes: int = 2,
                     use_class_weights: bool = False,
                     class_weights: Optional[torch.Tensor] = None,
                     focal_alpha: float = 1.0,
                     focal_gamma: float = 2.0,
                     device: torch.device = None) -> torch.nn.Module:
    """Get loss function based on name and parameters."""
    
    if loss_name.lower() == 'crossentropy':
        weights = class_weights if use_class_weights else None
        if weights is not None and device is not None:
            weights = weights.to(device)
        
        # Handle ignore_index parameter
        if ignore_index == False:
            ignore_idx = -100
        else:
            ignore_idx = ignore_index
            
        print(f"🎯 CrossEntropy Loss: Class weights {'ENABLED' if weights is not None else 'DISABLED'}")
        return torch.nn.CrossEntropyLoss(weight=weights, ignore_index=ignore_idx)
    
    elif loss_name.lower() == 'focal':
        weights = class_weights if use_class_weights else None
        if weights is not None and device is not None:
            weights = weights.to(device)
        print(f"🎯 Focal Loss: Alpha={focal_alpha}, Gamma={focal_gamma}, Class weights={'ENABLED' if weights is not None else 'DISABLED'}")
        return FocalLoss(alpha=focal_alpha, gamma=focal_gamma, ignore_index=ignore_index, weight=weights)
    
    else:
        raise ValueError(f"Unknown loss function: {loss_name}. "
                        f"Supported: 'crossentropy', 'focal'")

 2. Class Weight System
 
def compute_class_weights(labels_dir: str, 
                         num_classes: int,
                         ignore_index: int = -100,
                         custom_multipliers: Optional[Dict[int, float]] = None,
                         max_weight: float = 50.0,
                         use_inverse_frequency: bool = True) -> torch.Tensor:
    """
    Compute class weights for imbalanced datasets with optional custom multipliers and maximum weight cap.
    
    Args:
        labels_dir (str): Directory containing label files
        num_classes (int): Number of classes
        ignore_index (int): Index to ignore when computing weights
        custom_multipliers (dict, optional): Custom multipliers/weights for specific classes.
            - If use_inverse_frequency=True: Applied as multipliers after inverse frequency calculation
            - If use_inverse_frequency=False: Used as pure custom weights (all other classes get weight 1.0)
        max_weight (float): Maximum allowed weight value. Defaults to 50.0.
        use_inverse_frequency (bool): Whether to compute inverse frequency weights first. 
            - True (default): Compute inverse freq + apply custom multipliers
            - False: Use uniform weights (1.0) + apply custom weights only for specified classes
    """
    from collections import Counter
    import os
    
    # Count pixels for each class
    class_counts = Counter()
    total_pixels = 0
    
    # Get all label files
    label_extensions = (".tif", ".tiff", ".png", ".jpg", ".jpeg")
    label_files = [
        os.path.join(labels_dir, f)
        for f in os.listdir(labels_dir)
        if f.lower().endswith(label_extensions)
    ]
    
    print(f"Computing class weights from {len(label_files)} label files...")
    
    for label_file in label_files:
        try:
            if label_file.lower().endswith((".tif", ".tiff")):
                # Load GeoTIFF using rasterio
                import rasterio
                with rasterio.open(label_file) as src:
                    labels = src.read(1)
            else:
                # Load regular image using PIL
                from PIL import Image
                with Image.open(label_file) as img:
                    if img.mode != 'L':
                        img = img.convert('L')
                    labels = np.array(img)
            
            # Count valid pixels (not ignore_index)
            valid_pixels = labels[labels != ignore_index] if ignore_index >= 0 else labels
            pixel_counts = Counter(valid_pixels.flatten())
            
            for class_id, count in pixel_counts.items():
                class_counts[class_id] += count
                total_pixels += count
                
        except Exception as e:
            print(f"Warning: Could not process {label_file}: {e}")
            continue
    
    if total_pixels == 0:
        print("Warning: No valid pixels found. Using uniform weights.")
        return torch.ones(num_classes)
    
    # Initialize weights based on method choice
    weights = torch.ones(num_classes)
    
    if use_inverse_frequency:
        # Compute weights using inverse frequency
        for class_id in range(num_classes):
            # Skip weight computation for ignore_index
            if ignore_index >= 0 and class_id == ignore_index:
                weights[class_id] = 0.0
                continue
                
            if class_id in class_counts and class_counts[class_id] > 0:
                weights[class_id] = total_pixels / (num_classes * class_counts[class_id])
    else:
        # Use uniform weights (1.0 for all classes, 0.0 for ignore_index)
        for class_id in range(num_classes):
            if ignore_index >= 0 and class_id == ignore_index:
                weights[class_id] = 0.0
            else:
                weights[class_id] = 1.0

    # Apply custom multipliers/weights if provided
    if custom_multipliers:
        if use_inverse_frequency:
            print(f"\n🎯 Custom Class Weight Multipliers Configuration:")
            print(f"   ✅ Custom multipliers ENABLED: {custom_multipliers}")
            print("   📝 Applied as multipliers to inverse frequency weights")
        else:
            print(f"\n🎯 Pure Custom Class Weights Configuration:")
            print(f"   ✅ Custom weights ENABLED: {custom_multipliers}")
            print("   📝 Used as pure weights (no inverse frequency computation)")
        
        for class_id, multiplier in custom_multipliers.items():
            if 0 <= class_id < num_classes:
                original_weight = weights[class_id].item()
                
                if use_inverse_frequency:
                    # Apply as multiplier to existing inverse frequency weight
                    weights[class_id] *= multiplier
                else:
                    # Apply as pure custom weight (overwrites the default 1.0)
                    weights[class_id] = multiplier
                    
                new_weight = weights[class_id].item()
                print(f"   Class {class_id}: {original_weight:.4f} -> {new_weight:.4f}")

    # Apply maximum weight cap to prevent extreme values
    weights_capped = False
    print(f"\n🔒 Applying maximum weight cap of {max_weight}...")
    for class_id in range(num_classes):
        if weights[class_id] > max_weight:
            original_weight = weights[class_id].item()
            weights[class_id] = max_weight
            weights_capped = True
            print(f"   Class {class_id}: Capped weight from {original_weight:.4f} to {max_weight:.4f}")
    
    if not weights_capped:
        print(f"   ✅ No weights exceeded the maximum cap of {max_weight}")

    return weights

3. Enhanced Tile Filtering

def export_geotiff_tiles(
    input_raster: str,
    output_dir: str,
    input_labels: Optional[str] = None,
    tile_size: int = 256,
    stride: int = 128,
    skip_empty_tiles: bool = True,
    min_feature_ratio: Union[float, bool] = False,  # 🆕 NEW PARAMETER
    **kwargs
) -> List[str]:
    """
    Export GeoTIFF tiles with enhanced filtering capabilities.
    
    Args:
        min_feature_ratio (float or False): Minimum ratio of non-background pixels required in a tile.
            - False: Disable ratio filtering (original behavior)
            - 0.0-1.0: Minimum ratio of labeled pixels required to keep tile
            Only applies when skip_empty_tiles=True and input_labels is provided.
            
    Purpose: Skip tiles dominated by background (0-valued) pixels to improve training balance.
    """
    
    # ... existing tile generation logic ...
    
    # 🆕 ENHANCED FILTERING LOGIC
    if skip_empty_tiles and input_labels is not None:
        # Read label window
        with rasterio.open(input_labels) as label_src:
            label_window = label_src.read(1, window=window)
        
        # Check for completely empty tiles (original behavior)
        if np.all(label_window == 0):
            skipped_empty += 1
            if not (min_feature_ratio is False):
                # Only print if verbose mode or specific logging needed
                pass  # Reduced verbose logging
            continue
            
        # 🆕 Apply min_feature_ratio filtering if enabled
        if min_feature_ratio is not False:
            # Validate parameter
            if not isinstance(min_feature_ratio, (int, float)) or not (0.0 <= min_feature_ratio <= 1.0):
                raise ValueError("min_feature_ratio must be False or a float between 0.0 and 1.0")
            
            total_pixels = label_window.size
            feature_pixels = np.sum(label_window > 0)  # Non-background pixels
            feature_ratio = feature_pixels / total_pixels
            
            if feature_ratio < min_feature_ratio:
                skipped_background_heavy += 1
                # Reduced verbose logging - only count, don't print each tile
                continue
    
    # ... rest of tile generation logic ...
    
    # 🆕 ENHANCED SUMMARY REPORTING
    print(f"\n📊 Tile Generation Summary:")
    print(f"   ✅ Generated tiles: {tiles_created}")
    print(f"   🗑️  Skipped empty tiles: {skipped_empty}")
    if min_feature_ratio is not False:
        print(f"   🎯 Skipped background-heavy tiles: {skipped_background_heavy}")
        print(f"   📈 Feature ratio threshold: {min_feature_ratio:.1%}")
        print(f"   💪 Improved training balance by filtering {skipped_background_heavy + skipped_empty} low-content tiles")
    else:
        print(f"   ℹ️  Background ratio filtering: DISABLED")

# In the enhanced filtering section
if min_feature_ratio is not False:
    # Validate the parameter
    if not isinstance(min_feature_ratio, (int, float)) or not (0.0 <= min_feature_ratio <= 1.0):
        raise ValueError("min_feature_ratio must be False or a float between 0.0 and 1.0")
    
    # Calculate feature ratio
    total_pixels = label_window.size
    feature_pixels = np.sum(label_window > 0)  # Count non-background pixels
    feature_ratio = feature_pixels / total_pixels
    
    # Filter out background-heavy tiles
    if feature_ratio < min_feature_ratio:
        skipped_background_heavy += 1
        continue  # Skip this tile

4. Flexible Ignore Index

def get_loss_function(loss_name: str, 
                     ignore_index: Union[int, bool] = -100,  # 🆕 FLEXIBLE TYPE
                     num_classes: int = 2,
                     use_class_weights: bool = False,
                     class_weights: Optional[torch.Tensor] = None,
                     focal_alpha: float = 1.0,
                     focal_gamma: float = 2.0,
                     device: torch.device = None) -> torch.nn.Module:
    """
    Args:
        ignore_index (int or bool): Index to ignore in loss calculation, or False to disable ignoring.
            - If int: pixels with this label value will be ignored during training
            - If False: no pixels will be ignored (all pixels contribute to loss)
    """
    
    if loss_name.lower() == 'crossentropy':
        # 🆕 HANDLE FLEXIBLE IGNORE INDEX
        if ignore_index == False:
            # Don't ignore any index - use a value that won't match any target
            ignore_idx = -100
        else:
            # Use the specified ignore_index (integer)
            ignore_idx = ignore_index
            
        return torch.nn.CrossEntropyLoss(weight=weights, ignore_index=ignore_idx)
    
    elif loss_name.lower() == 'focal':
        # 🆕 PASS FLEXIBLE IGNORE INDEX TO FOCAL LOSS
        return FocalLoss(alpha=focal_alpha, gamma=focal_gamma, ignore_index=ignore_index, weight=weights)
```

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

Implements the enhancements from #335 to support training with discrete class landcover data.

### Changes

**New in `train.py`:**
- **`FocalLoss`** class — focal loss for class imbalance (`FL = -α(1-pt)^γ log(pt)`), with configurable alpha, gamma, ignore_index (int or `False` to disable), reduction, and per-class weights. Uses `F.log_softmax` + `F.nll_loss` internally.
- **`get_loss_function()`** — factory returning configured `CrossEntropyLoss` or `FocalLoss` with flexible `ignore_index` and optional class weights.
- **`compute_class_weights()`** — computes per-class weights from label tiles with inverse-frequency mode, custom multipliers, and weight capping.
- **`train_segmentation_model()`** — wired with new parameters: `loss_function`, `ignore_index`, `use_class_weights`, `class_weights`, `custom_class_multipliers`, `max_class_weight`, `use_inverse_frequency`, `focal_alpha`, `focal_gamma`.

**New in `utils.py`:**
- **`min_feature_ratio`** parameter on `export_geotiff_tiles()` — filters out tiles where the ratio of non-background pixels is below a threshold (only applies when `skip_empty_tiles=True` and label data is provided). Summary stats are reported at the end.

**New in `__init__.py`:**
- Exports `FocalLoss`, `get_loss_function`, `compute_class_weights` from `train.py`
- Exports `LandcoverCrossEntropyLoss`, `landcover_iou`, `get_landcover_loss_function`, `train_segmentation_landcover` from `landcover_train.py`
- Exports `export_landcover_tiles` from `landcover_utils.py`

### Backward Compatibility
All new parameters have sensible defaults — existing code continues to work without changes.

Closes #335

---
Closes #335

### CODE REVIEW

**Comment by gw0ods** [comment]
Hello Professor Wu, I tried reinstalling geoai on another system and the modules I added didnt seem to work properly... I had to change a few things in them. Sorry about that, I cant push a working version next week I think.
I also added a better IOU method for this discrete landcover classification. The issue was that when there is sparse data false positives shouldnt be considered in IOU calculation and model selection.

**Comment by giswqs** [comment]
@gw0ods Can you open a new issue and describe what is not working?

**Comment by gw0ods** [comment]
> @gw0ods Can you open a new issue and describe what is not working?

on it!
